File: lua/queen-attack/queen-attack.lua
return function(pos)
  assert(pos.row >= 0 and pos.row < 8, 'row from 0 to 7')
  assert(pos.column >= 0 and pos.column < 8, 'column from 0 to 7')
  return {
    row = pos.row,
    column = pos.column,
    can_attack = function(other)
      local dy = math.abs(pos.row - other.row)
      local dx = math.abs(pos.column - other.column)
      return dx == 0 or dy == 0 or dx == dy
    end
  }
end

File: zig/queen-attack/queen_attack.zig
const absCast = @import("std").math.absCast;

pub const QueenError = error{
    InitializationFailure,
};
pub const Queen = struct {
    x: i8,
    y: i8,

    pub fn init(row: i8, column: i8) QueenError!Queen {
        if (row >= 0 and row < 8 and column >= 0 and column < 8) {
            return Queen{ .x = row, .y = column };
        }
        return QueenError.InitializationFailure;
    }

    pub fn canAttack(self: *const Queen, other: Queen) !bool {
        const dx = absCast(self.x - other.x);
        const dy = absCast(self.y - other.y);
        return dx == 0 or dy == 0 or dx == dy;
    }
};
